sql/src/main/java/io/crate/analyze/AnalyzedInsertStatement.java

```diff
@@ -34,18 +35,37 @@
 
     private final DocTableInfo targetTable;
     private final List<Reference> targetCols;
-    private final List<List<Symbol>> rows;
+    @Nullable
+    private List<List<Symbol>> rows;
+    @Nullable
+    private AnalyzedStatement analyzedRelation;
     private final Map<Reference, Symbol> onDuplicateKeyAssignments;
 
+    private AnalyzedInsertStatement(DocTableInfo targetTable,
+                                    List<Reference> targetCols,
+                                    Map<Reference, Symbol> onDuplicateKeyAssignments) {
+
+        this.targetTable = targetTable;
+        this.targetCols = targetCols;
+        this.onDuplicateKeyAssignments = onDuplicateKeyAssignments;
+    }
+
     public AnalyzedInsertStatement(DocTableInfo targetTable,
                                    List<Reference> targetCols,
                                    List<List<Symbol>> rows,
                                    Map<Reference, Symbol> onDuplicateKeyAssignments) {
 
-        this.targetTable = targetTable;
-        this.targetCols = targetCols;
+        this(targetTable, targetCols, onDuplicateKeyAssignments);
         this.rows = rows;
-        this.onDuplicateKeyAssignments = onDuplicateKeyAssignments;
+    }
+
+    public AnalyzedInsertStatement(DocTableInfo targetTable,
+                                   List<Reference> targetCols,
+                                   AnalyzedStatement analyzedRelation,
```

Comment:
I'd rename the parameter and the attribute to subRelation, it's a bit more specific.